Avoid double decoding all client responses

smarterclayton opened the pull request:
Fixes #35982 

The linked issue uncovered that we were always double decoding the response in restclient for get, list, update, create, and patch.  That's fairly expensive, most especially for list.  This PR refines the behavior of the rest client to avoid double decoding, and does so while minimizing the changes to rest client consumers.

restclient must be able to deal with multiple types of servers. Alter the behavior of restclient.Result#Raw() to not process the body on error, but instead to return the generic error (which still matches the error checking cases in api/error like IsBadRequest). If the caller uses
.Error(), .Into(), or .Get(), try decoding the body as a Status.

For older servers, continue to default apiVersion "v1" when calling restclient.Result#Error(). This was only for 1.1 servers and the extensions group, which we have since fixed.

This removes a double decode of very large objects (like LIST) - we were trying to DecodeInto status, but that ends up decoding the entire result and then throwing it away.  This makes the decode behavior specific to the type of action the user wants.

```release-note
The error handling behavior of `pkg/client/restclient.Result` has changed.  Calls to `Result.Raw()` will no longer parse the body, although they will still return errors that react to `pkg/api/errors.Is*()` as in previous releases.  Callers of `Get()` and `Into()` will continue to receive errors that are parsed from the body if the kind and apiVersion of the body match the `Status` object.

This more closely aligns rest client as a generic RESTful client, while preserving the special Kube API extended error handling for the `Get` and `Into` methods (which most Kube clients use).
```

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36001)
<!-- Reviewable:end -->

smarterclayton committed:
Avoid double decoding all client responses

restclient must be able to deal with multiple types of servers. Alter
the behavior of restclient.Result#Raw() to not process the body on
error, but instead to return the generic error (which still matches the
error checking cases in api/error like IsBadRequest). If the caller uses
.Error(), .Into(), or .Get(), try decoding the body as a Status.

For older servers, continue to default apiVersion "v1" when calling
restclient.Result#Error(). This was only for 1.1 servers and the
extensions group, which we have since fixed.

This removes a double decode of very large objects (like LIST).

smarterclayton commented:
@kubernetes/sig-api-machinery @wojtek-t this doubles client read performance (even versioned clients).

smarterclayton commented:
I added a release note that describes how error handling has changed - all calls to `errors.IsNotFound/*` should be the same for `Raw`, but the body will no longer be parsed unless clients explicitly ask for it.  I did a quick review of all the places where we call Raw and didn't find any places that could be broken by this (/healthz, /version, /metrics, /swagger, and test metrics loads)

caesarxuchao commented on pkg/client/restclient/request.go:
```diff
@@ -1101,7 +1134,29 @@ func (r Result) WasCreated(wasCreated *bool) Result {
 func (r Result) Error() error {
+	// if we have received an unexpected server error, and we have a body and decoder, we can try to extract
+	// a Status object.
+	if r.err == nil || !errors.IsUnexpectedServerError(r.err) || len(r.body) == 0 || r.decoder == nil {
```
Why `IsUnexpectedServerError`? I think we should just check if `r.err` implements APIStatus interface.

smarterclayton replied:
No, the error in this case is set by the thing that creates Result (which calls newGeneric...Error, which is UnexpectedServerError).

UnexpectedServerError literally means - we got an error from the server but we couldn't / didn't decode the body. So this code is looking for errors where before we didn't try to decode it, and try to decode it.

caesarxuchao commented on pkg/client/restclient/request_test.go:
```diff
@@ -626,9 +628,46 @@ func TestTransformUnstructuredError(t *testing.T) {
+			},
+		},
+		{
+			// successful status is ignored
```
When will this happen in real life?

caesarxuchao commented:
LGTM. @deads2k if you want to take another look.

smarterclayton commented:
Found a bug - authorization didn't register Status.  Is authorization new for 1.5 or did we also fail to register it in 1.4?

liggitt commented:
Was in 1.4

smarterclayton commented:
I'm going to fix it in 1.5 but that means 1.4 clients can't get Status messages from subject access review.  Was it alpha?

smarterclayton committed:
ExportOptions should be registered to all public schemas

smarterclayton committed:
Improve the common kinds API test to catch more errors

smarterclayton committed:
Authorization should be registered in testapi

smarterclayton committed:
Unversioned types should decode no matter what the requested version is

Status is exposed as v1 in the current schema (so all groups are
returning v1.Status).  However, if you give a codec only "mygroup"
"myversion", it will fail to convert Status to v1. For now, unversioned
types should be allowed to be projected into all group versions, and
when we add the server group we'll rip out the unversioned concept
entirely.

smarterclayton commented:
Ok, that was fun.  

1. ExportOptions wasn't registered in all groups - fixed that.
2. Added a test to verify ExportOptions was in all groups and could decode / encode correctly
3. Authorization wasn't registered in testapi
4. Found a bug with how we treat Status - it's projected into all groups as v1, when it really should be the group version.  If you called runtime.ConvertToVersion with "your group version" it would fail to convert a v1.Status response into unversioned.Status.  Added a case to convertToVersion to bandaid over this until I add the server api group version and rip out unversioned.

Ready for re-review.  No backwards compatibility impact now.

deads2k commented on pkg/api/testapi/testapi.go:
```diff
@@ -279,6 +290,7 @@ func init() {
 	Rbac = Groups[rbac.GroupName]
 	Storage = Groups[storage.GroupName]
 	ImagePolicy = Groups[imagepolicy.GroupName]
+	Authorization = Groups[authorization.GroupName]
```
If nobody uses this, don't add it.  We want less code to touch for each group you add, not more.

deads2k commented on pkg/api/testapi/testapi.go:
```diff
@@ -258,6 +260,15 @@ func init() {
 			externalTypes:        api.Scheme.KnownTypes(externalGroupVersion),
 		}
 	}
+	if _, ok := Groups[authorization.GroupName]; !ok {
```
It'd be nice to have a unit test somewhere that makes sure this parallel registration stays in sync.

deads2k commented:
nits, lgtm.

smarterclayton commented:
I need it to properly test authorization implements ExportOptions, which is
what this issue uncovered.

On Wed, Nov 2, 2016 at 7:53 AM, David Eads <notifications@github.com> wrote:

> *@deads2k* commented on this pull request.
> ------------------------------
>
> In pkg/api/testapi/testapi.go
> <https://github.com/kubernetes/kubernetes/pull/36001#pullrequestreview-6797347>
> :
>
> > @@ -279,6 +290,7 @@ func init() {
>  	Rbac = Groups[rbac.GroupName]
>  	Storage = Groups[storage.GroupName]
>  	ImagePolicy = Groups[imagepolicy.GroupName]
> +	Authorization = Groups[authorization.GroupName]
>
> If nobody uses this, don't add it. We want less code to touch for each
> group you add, not more.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36001#pullrequestreview-6797347>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p89wRSx18Qsc9yIqaCAFfpyZ5kZBks5q6Hm2gaJpZM4Kmjnm>
> .
>

smarterclayton commented:
I agree.  Do you have an issue that covers your concerns with testapi that
I can add to?

On Wed, Nov 2, 2016 at 7:54 AM, David Eads <notifications@github.com> wrote:

> *@deads2k* commented on this pull request.
> ------------------------------
>
> In pkg/api/testapi/testapi.go
> <https://github.com/kubernetes/kubernetes/pull/36001#pullrequestreview-6797441>
> :
>
> > @@ -258,6 +260,15 @@ func init() {
>  			externalTypes:        api.Scheme.KnownTypes(externalGroupVersion),
>  		}
>  	}
> +	if _, ok := Groups[authorization.GroupName]; !ok {
>
> It'd be nice to have a unit test somewhere that makes sure this parallel
> registration stays in sync.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36001#pullrequestreview-6797441>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3Ovqif8TFW6SW325KaNQWv3qq1wks5q6HncgaJpZM4Kmjnm>
> .
>

smarterclayton commented:
As per comments

/lgtm

On Wed, Nov 2, 2016 at 11:48 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I agree.  Do you have an issue that covers your concerns with testapi that
> I can add to?
>
> On Wed, Nov 2, 2016 at 7:54 AM, David Eads <notifications@github.com>
> wrote:
>
>> *@deads2k* commented on this pull request.
>> ------------------------------
>>
>> In pkg/api/testapi/testapi.go
>> <https://github.com/kubernetes/kubernetes/pull/36001#pullrequestreview-6797441>
>> :
>>
>> > @@ -258,6 +260,15 @@ func init() {
>>  			externalTypes:        api.Scheme.KnownTypes(externalGroupVersion),
>>  		}
>>  	}
>> +	if _, ok := Groups[authorization.GroupName]; !ok {
>>
>> It'd be nice to have a unit test somewhere that makes sure this parallel
>> registration stays in sync.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/36001#pullrequestreview-6797441>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p3Ovqif8TFW6SW325KaNQWv3qq1wks5q6HncgaJpZM4Kmjnm>
>> .
>>
>
>

deads2k commented:
> I agree.  Do you have an issue that covers your concerns with testapi that
> I can add to?

Old PR trying to remove the package entirely: https://github.com/kubernetes/kubernetes/pull/34644 .  I'm still not sure that I'm convinced of the value it adds.  I thought new groups would be able to ignore the package, so I didn't press on it.

caesarxuchao commented:
> ExportOptions wasn't registered in all groups - fixed that.

Can you add ExportOptions to this [line](https://github.com/kubernetes/kubernetes/pull/27733/files#diff-394377845c8dfd9eec363c059cdde354R177) to test it's registered in all groups?

caesarxuchao commented:
Sorry, I saw you did that :)

caesarxuchao commented:
Fix https://github.com/kubernetes/kubernetes/issues/27886.
